recitation_2/makefile.md

```diff
@@ -0,0 +1,401 @@
+
+- Replace `C_FILES = hello.c goodbye.c myprogram.c` with `C_FILES = $(wildcard *.c)`
+
+**`patsubst` Function:**
+
+`OBJS` are all the `.c` files in our current directory, but with a `.o` extension instead of `.c`. We can use the `patsubst` function and pattern matching instead. `$(patsubst %.c,%.o,$(C_FILES))` takes all the file names in `C_FILES` and replaces `.c` with `.o`.
+
+- Replace `OBJS = hello.o goodbye.o myprogram.o` with `OBJS = $(patsubst %.c,%.o,$(C_FILES))`
+
+**Special Macros:**
+
+For our rules for `.o` targets, the output name of the recipe (what follows the `-o` flag) is always the same as the name of the target and the ingredient is always its corresponding `.c` file, which is the first dependency listed in our rule. We can use special macros to generalize this.
+
+`$@` evaluates to the name of the current target.
+
+`$<` evaluates to the first dependency.
+
+- Replace `$(CC) -c -o <output name> <ingredient>` with `$(CC) $(CFLAGS) -c -o $@ $<`
+
+**`%` (Pattern Matching):**
+
+Our rules for `.o` targets all have dependencies for the corresponding `.c` and `.h` files (with the same file name before the extension). We can generalize this with pattern matching. `%.o: %.c %.h` means for every `.o` file that exists, its dependencies are the `.c` and `.h` files of the same file name (before the file extension).
+
+- Replace `<filename>.o: <filename>.c <filename>.h` with `%.o: %.c %.h`
+
+
+
+Altogether, we get this `Makefile`:
+
+```bash
+CC      = gcc
+TARGET  = myprogram
+C_FILES = $(wildcard *.c)
+OBJS    = $(patsubst %.c,%.o,$(C_FILES))
+
+$(TARGET): $(OBJS)
+	$(CC) -o $(TARGET) $(OBJS)
+    
+%.o: %.c %.h
+	$(CC) -c -o $@ $<
+```
+
+
+
+However, note that  `myprogram.o` doesn’t depend on `myprogram.h`, as no such file exists. 
+
+So, we can filter out `myprogram.c` from our `C_FILES` with the **`filter-out`** function.
+
+- Replace `C_FILES = $(wildcard *.c)` with `C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c)) `
+
+We then have to compile and link `myprogram.c` with the rest of the object files to produce `myprogram`, so we can change our rule for `myprogram` to:
+
+```bash
+$(TARGET): $(OBJS) $(TARGET).c
+	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c
+```
+
+This works because the command `gcc -o <output name> <ingredients>` will both compile (if needed for `.c` files) and link your ingredients.
+
+Note: For multi-source programs, however, it is not good practice to compile and link all your files in one step because it defeats the purpose of incremental building and individual dependencies cannot be checked. In our `Makefile`, aside from the `.c` file containing the main function, all other `.o` objects are still built separately.
```

Comment:
```suggestion
Note: For multi-source programs, however, it is not good practice to compile and link all your files in one step because it defeats the purpose of incremental building, and individual dependencies cannot be checked. In our `Makefile`, aside from the `.c` file containing the main function, all other `.o` objects are still built separately.
```

Reply:
fanboys